client/gcp/gcp.go

```diff
@@ -0,0 +1,212 @@
+	if nextpageToken != "" {
+		endpoint = fmt.Sprintf("%v&pageToken=%v", PriceInfoEndpoint, nextpageToken)
+	}
+
+	req, err := http.NewRequest("GET", endpoint, nil)
+	req.Header.Add("User-Agaent", `Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.4 Safari/605.1.15`)
+	req.Header.Add("Connection", `keep-alive`)
+	req.Header.Add("Accept", `text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8`)
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Fail to fetch the info file, [internal]: %v", err)
+	}
+	if res.StatusCode != 200 {
+		return nil, fmt.Errorf("An http error occur , [internal]: %v", err)
+	}
+
+	data, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Fail when reading response data , [internal]: %v", err)
+	}
+	var offerList []*offer
+	rawData := &pricing{}
+	if err := json.Unmarshal(data, rawData); err != nil {
+		return nil, fmt.Errorf("Fail when unmarshaling response data , [internal]: %v", err)
+	}
+	offerList = rawData.OfferList
+
+	if rawData.NextPageToken != "" {
+		nextPage, err := httpHelper(rawData.NextPageToken)
+		if err != nil {
+			return nil, err
+		}
+		offerList = append(offerList, nextPage...)
+		return offerList, nil
+	}
+
+	return offerList, nil
+}
+
+// GetOffer return the offers provide by GCP.
+func (c *Client) GetOffer() ([]*client.Offer, error) {
+	rawOffer, err := httpHelper("")
+	if err != nil {
+		return nil, err
+	}
+
+	var offerList []*client.Offer
+	incrID := 0
+	for _, rawOffer := range rawOffer {
+		// This condition will filter resource like Network.
+		if rawOffer.Category.ResourceFamily != "ApplicationServices" ||
+			// the Gen1 offering is unavailabe
+			strings.Contains(rawOffer.Category.ResourceGroup, "Gen1") ||
+			!rSpecification.Match([]byte(rawOffer.Description)) {
+			continue
+		}
+
+		offerType := getOfferType(rawOffer.Category.ResourceGroup)
+		hourlyUSD := getUSD(rawOffer.PricingInfo[0].PricingExpression.TieredRateList[0].UnitPrice)
+
+		offer := &client.Offer{
+			ID:        incrID,
+			SKU:       rawOffer.ID,
+			OfferType: offerType,
+
+			// All GCP services are charged on demand literal, but we consider commitment for a length as reserved type
+			ChargeType: client.ChargeTypeOnDemand,
+
+			RegionList:  rawOffer.ServiceRegionList,
+			Description: rawOffer.Description,
+
+			CommitmentUSD: 0,
+			HourlyUSD:     hourlyUSD,
+		}
+		incrID++
+
+		if offerType == client.OfferTypeInstance {
+			databaseEngine, CPU, memory := getCPUMemory(rawOffer.Description)
+			instanceType := getInstanceType(rawOffer.Category.ResourceGroup, CPU, memory)
+			payload := &client.OfferInstancePayload{
+				Type:           instanceType,
+				InstanceFamily: instanceType,
+				CPU:            CPU,
+				Memory:         memory,
+				DatabaseEngine: databaseEngine,
+			}
+			offer.InstancePayload = payload
+		}
+
+		if offer.ChargeType == client.ChargeTypeReserved {
+			panic("Implement me")
```

Comment:
Don't panic :-) try something else such as "TODO(zilong): implement it later." then continue.